src/hotspot/share/nmt/memLogRecorder.cpp

```diff
@@ -0,0 +1,839 @@
+  int     fd;
+} file_info;
+
+static file_info _open_file_and_read(const char* pattern, const char* path, int pid) {
+  file_info info = { nullptr, 0, -1 };
+
+  char *file_path = NEW_C_HEAP_ARRAY(char, JVM_MAXPATHLEN, mtNMT);
+  if (!_create_file_path_with_pid(path, pattern, file_path, pid)) {
+    tty->print("Can't construct path [%s:%s:%d].", pattern, path, pid);
+    return info;
+  }
+
+  info.fd = os::open(file_path, O_RDONLY, 0);
+  if (info.fd == -1) {
+    int e = errno;
+    tty->print("Can't open file [%s].", file_path);
+    tty->print_raw_cr(os::strerror(e));
+    return info;
+  }
+
+  struct stat file_info;
+  ::fstat(info.fd, &file_info);
+  info.size = file_info.st_size;
+  ::lseek(info.fd, 0, SEEK_SET);
+
+#if !defined(_WIN64)
+  info.ptr = ::mmap(nullptr, info.size, PROT_READ, MAP_PRIVATE, info.fd, 0);
+  assert(info.ptr != MAP_FAILED, "info.ptr != MAP_FAILED");
+#endif
+
+  FREE_C_HEAP_ARRAY(char, file_path);
+
+  return info;
+}
+
+void NMT_MemoryLogRecorder::initialize(long int limit) {
+  // tty->print("> NMT_MemoryLogRecorder::initialize(%ld, %ld)\n", limit, sizeof(Entry));
+  NMT_MemoryLogRecorder *recorder = NMT_MemoryLogRecorder::instance();
+  recorder->init();
+  NMTRecorder_Locker locker;
+  {
+    recorder->_limit = limit;
+    if (recorder->_limit > 0) {
+      recorder->_log_fd = _prepare_log_file(nullptr, ALLOCS_LOG_FILE);
+      recorder->_done = false;
+    } else {
+      recorder->_done = true;
+    }
+  }
+}
+
+void NMT_MemoryLogRecorder::finish(void) {
+  NMT_MemoryLogRecorder *recorder = NMT_MemoryLogRecorder::instance();
+  // tty->print("NMT_MemoryLogRecorder::finish() %p\n", NMT_MemoryLogRecorder::instance());
+  if (!recorder->done()) {
+    NMTRecorder_Locker locker;
+
+    volatile int log_fd = recorder->_log_fd;
+    recorder->_log_fd = -1;
+    // tty->print(" log_fd:%d\n", log_fd);
+    log_fd = _close_and_check(log_fd);
+    // tty->print(" log_fd:%d\n", log_fd);
+
+    int threads_fd = _prepare_log_file(nullptr, THREADS_LOG_FILE);
+    // tty->print(" threads_fd:%d\n", threads_fd);
+    if (threads_fd != -1) {
+      _write_and_check(threads_fd, recorder->_threads_names, (recorder->_threads_names_size-1)*sizeof(thread_name_info));
+      threads_fd = _close_and_check(threads_fd);
+      // tty->print(" threads_fd:%d\n", threads_fd);
+    }
+
+    int info_fd = _prepare_log_file(nullptr, INFO_LOG_FILE);
+    // tty->print(" info_fd:%d\n", info_fd);
+    if (info_fd != -1) {
+      size_t level = NMTUtil::parse_tracking_level(NativeMemoryTracking);
+      _write_and_check(info_fd, &level, sizeof(level));
+      size_t overhead = MemTracker::overhead_per_malloc();
+      _write_and_check(info_fd, &overhead, sizeof(overhead));
+      info_fd = _close_and_check(info_fd);
+      // tty->print(" info_fd:%d\n", info_fd);
+    }
+
+    recorder->_done = true;
+  }
+  os::exit(0);
+}
+
+void NMT_MemoryLogRecorder::replay(const int pid) {
+  // tty->print("NMT_MemoryLogRecorder::replay(%d)\n", pid);
```

Comment:
Please make sure that you've removed all commented out prints if you don't intend to keep them.